File: PseudoCoupHQ/Research/oracle/riscv/softfloat_slices/emulations/rust/f32_roundToInt.rs
#![allow(non_snake_case, unused_parens, unused_imports, unused_variables, unused_comparisons, clippy::all)]
use crate::helpers::*;


pub fn f32_roundToInt_rm1(v_arg: u64, v_arg1: bool) -> u64 {
    let v_i: u32 = ((v_arg) as u32);
    let v_i2: u32 = v_i.wrapping_shr((0x17u32) as u32);
    let v_i3: u32 = (v_i2 & 0xffu32);
    let v_i4: bool = (v_i3 < 0x7fu32);
    let v_i6: u32 = (v_i & 0x7fffffffu32);
    let v_i7: bool = (v_i6 == 0x0u32);
    let v__m1: u32 = (0u32.wrapping_sub((v_i7) as u32));
    let v_i8: u32 = (v_i & 0x80000000u32);
    let v__a2: u32 = (v_i & v__m1);
    let v__n3: u32 = (v__m1 ^ 0xffffffffu32);
    let v__a4: u32 = (v_i8 & v__n3);
    let v_spec_select: u32 = (v__a2 | v__a4);
    let v_i10: bool = (v_i3 > 0x95u32);
    let v_i12: bool = (v_i3 != 0xffu32);
    let v_i17: u32 = 0x96u32.wrapping_sub(v_i3);
    let v__sh9: u32 = 0xffffffffu32.wrapping_shl((v_i17) as u32);
    let v__neg: u32 = v__sh9;
    let v_i18: u32 = (v__neg & v_i);
    let v_i13: u32 = (v_i & 0x7fffffu32);
    let v_i14: bool = (v_i13 == 0x0u32);
    let v_i15: bool = (v_i14 | v_i12);
    let v__m5: u32 = (0u32.wrapping_sub((v_i15) as u32));
    let v__a6: u32 = (v_i & v__m5);
    let v__n7: u32 = (v__m5 ^ 0xffffffffu32);
    let v__a8: u32 = (0x7fc00000u32 & v__n7);
    let v_spec_select4: u32 = (v__a6 | v__a8);
    let v__n10: bool = (v_i10 ^ true);
    let v__n11: bool = (v_i4 ^ true);
    let v__c12: bool = (v__n10 & v__n11);
    let v__m13: u32 = (0u32.wrapping_sub((v__c12) as u32));
    let v__a14: u32 = (v_i18 & v__m13);
    let v__m15: u32 = (0u32.wrapping_sub((v_i4) as u32));
    let v__a16: u32 = (v_spec_select & v__m15);
    let v__o17: u32 = (v__a14 | v__a16);
    let v__m19: u32 = (0u32.wrapping_sub((v_i10) as u32));
    let v__a20: u32 = (v_spec_select4 & v__m19);
    let v__o21: u32 = (v__o17 | v__a20);
    let v_i20: u64 = ((v__o21) as u64);
    v_i20
}

File: PseudoCoupHQ/Research/oracle/riscv/softfloat_slices/emulations/js/f32_roundToInt.js
'use strict';
const { sfUdiv, sfSgn, sfCtlz, sfAbs, sfUsubsat, sfFshl } = require('./helpers.js');


function f32_roundToInt_rm1(v_arg, v_arg1) {
  const v_i = (v_arg & 0xffffffffn);
  const v_i2 = (v_i >> (0x17n & 0x1fn));
  const v_i3 = (v_i2 & 0xffn);
  const v_i4 = ((v_i3 < 0x7fn) ? 1n : 0n);
  const v_i6 = (v_i & 0x7fffffffn);
  const v_i7 = ((v_i6 === 0x0n) ? 1n : 0n);
  const v__m1 = (sfSgn(v_i7, 0x1n) & 0xffffffffn);
  const v_i8 = (v_i & 0x80000000n);
  const v__a2 = (v_i & v__m1);
  const v__n3 = (v__m1 ^ 0xffffffffn);
  const v__a4 = (v_i8 & v__n3);
  const v_spec_select = (v__a2 | v__a4);
  const v_i10 = ((v_i3 > 0x95n) ? 1n : 0n);
  const v_i12 = ((v_i3 !== 0xffn) ? 1n : 0n);
  const v_i17 = ((0x96n - v_i3) & 0xffffffffn);
  const v__sh9 = ((0xffffffffn << (v_i17 & 0x1fn)) & 0xffffffffn);
  const v__neg = v__sh9;
  const v_i18 = (v__neg & v_i);
  const v_i13 = (v_i & 0x7fffffn);
  const v_i14 = ((v_i13 === 0x0n) ? 1n : 0n);
  const v_i15 = (v_i14 | v_i12);
  const v__m5 = (sfSgn(v_i15, 0x1n) & 0xffffffffn);
  const v__a6 = (v_i & v__m5);
  const v__n7 = (v__m5 ^ 0xffffffffn);
  const v__a8 = (0x7fc00000n & v__n7);
  const v_spec_select4 = (v__a6 | v__a8);
  const v__n10 = (v_i10 ^ 0x1n);
  const v__n11 = (v_i4 ^ 0x1n);
  const v__c12 = (v__n10 & v__n11);
  const v__m13 = (sfSgn(v__c12, 0x1n) & 0xffffffffn);
  const v__a14 = (v_i18 & v__m13);
  const v__m15 = (sfSgn(v_i4, 0x1n) & 0xffffffffn);
  const v__a16 = (v_spec_select & v__m15);
  const v__o17 = (v__a14 | v__a16);
  const v__m19 = (sfSgn(v_i10, 0x1n) & 0xffffffffn);
  const v__a20 = (v_spec_select4 & v__m19);
  const v__o21 = (v__o17 | v__a20);
  const v_i20 = (v__o21);
  return v_i20;
}
module.exports = { f32_roundToInt_rm1 };
